### PR TITLE
Introduce `DatabricksEnvironment` and fix Azure MSI auth from ACR, where IMDS doesn't give host environment information

### PR DESCRIPTION
**Pull request by nfx**
## Changes
This PR allows determining Azure Environment from a Databricks account or workspace hostname, removing the need for a separate configuration/environment variable and complexities related to Azure MSI from within ACR.

Similar functionality in Python SDK: 
- https://github.com/databricks/databricks-sdk-py/pull/390

Stacked on top of:
- https://github.com/databricks/databricks-sdk-go/pull/699

## Tests

- [x] `make test` passing
- [x] `make fmt` applied
- [ ] relevant integration tests applied

### CODE REVIEW

File: config/environments.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+func (c *Config) Environment() (*DatabricksEnvironment, error) {
+	hostname := c.CanonicalHostName()
+	for _, e := range append(c.DatabricksEnvironments, envs...) {
+		if strings.HasSuffix(hostname, e.DnsZone) {
+			return &e, nil
+		}
+	}
+	// TODO: do we return default one or do we return error?...
+	return nil, fmt.Errorf("cannot find DatabricksEnvironment for %s", hostname)
```
Comment (mgyucht) [inline config/environments.go:59]: Let's default to AWS Prod environment since this case will be hit for PVC.

File: config/environments.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	CloudUnspecified Cloud = "Unspecified"
+	CloudAWS         Cloud = "AWS"
+	CloudAzure       Cloud = "Azure"
+	CloudGCP         Cloud = "GCP"
+)
+
+type DatabricksEnvironment struct {
+	Cloud Cloud
+
+	// zones are not very secret: https://crt.sh/?q=databricks
```
Comment (mgyucht) [inline config/environments.go:20]: You can remove this comment

File: config/environments.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	DnsZone string
+
+	// The application (client) ID isn't a secret.
```
Comment (mgyucht) [inline config/environments.go:23]: You can remove this comment.

File: config/environments.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+	{Cloud: CloudAWS, DnsZone: ".dev.databricks.com"},
+	{Cloud: CloudAWS, DnsZone: ".staging.cloud.databricks.com"},
+	{Cloud: CloudAWS, DnsZone: ".cloud.databricks.com"},
+	{Cloud: CloudAWS, DnsZone: ".cloud.databricks.us"},
+
+	{Cloud: CloudAzure, DnsZone: ".dev.azuredatabricks.net", AzureApplicationID: "62a912ac-b58e-4c1d-89ea-b2dbfc7358fc", AzureEnvironment: &publicCloud},
+	{Cloud: CloudAzure, DnsZone: ".staging.azuredatabricks.net", AzureApplicationID: "4a67d088-db5c-48f1-9ff2-0aace800ae68", AzureEnvironment: &publicCloud},
+	{Cloud: CloudAzure, DnsZone: ".azuredatabricks.net", AzureApplicationID: "2ff814a6-3304-4ab8-85cb-cd0e6f879c1d", AzureEnvironment: &publicCloud},
+	{Cloud: CloudAzure, DnsZone: ".databricks.azure.us", AzureApplicationID: "2ff814a6-3304-4ab8-85cb-cd0e6f879c1d", AzureEnvironment: &usGovernmentCloud},
```
Comment (mgyucht) [inline config/environments.go:44]: Can we add the China endpoints here?

File: config/config.go [hunk 5]
```diff
@@ -139,16 +142,23 @@ func (c *Config) Authenticate(r *http.Request) error {
 
 // IsAzure returns if the client is configured for Azure Databricks.
 func (c *Config) IsAzure() bool {
-	isAzureHost := strings.Contains(c.Host, ".azuredatabricks.net") ||
-		strings.Contains(c.Host, "databricks.azure.cn") ||
-		strings.Contains(c.Host, ".databricks.azure.us")
-
-	return isAzureHost || c.AzureResourceID != ""
+	if c.AzureResourceID != "" {
```
Comment (mgyucht) [inline config/config.go:145]: These are much nicer :)

File: config/config.go [hunk 6]
```diff
@@ -70,10 +70,10 @@ type Config struct {
 	AzureClientID     string `name:"azure_client_id" env:"ARM_CLIENT_ID" auth:"azure"`
 	AzureTenantID     string `name:"azure_tenant_id" env:"ARM_TENANT_ID" auth:"azure"`
 
-	// AzureEnvironment (Public, UsGov, China, Germany) has specific set of API endpoints.
+	// [Deprecated] AzureEnvironment (Public, UsGov, China, Germany) has specific set of API endpoints.
```
Comment (mgyucht) [inline config/config.go:73]: Beyond saying "deprecated", let's also say that they no longer have any effect.

**Review by mgyucht — COMMENTED** [review]
Couple small suggestions, but mostly looking good. Thanks!

File: config/config.go (hunk 6 above)
Comment (nfx) [reply config/config.go:73]: ah wait, we do still need AzureEnvironment for cases when no `host` and only `azure_workspace_resource_id` is supplied, we have to call resource management to fetch the workspace url.

**Pull request merged**